README Re-write (#178)
kmagiera/react-native-reanimated: README.md

React Native's Animated library reimplemented.

- **Native Performance**: Declare your animations in JS, but have them run on the native thread! 🧙‍♂️
- **Precise Animations**: The API affords new levels of precision and detailed control of your animations. 🕹
- **Backwards Compatible**: Use the same Animated library API from React Native that you've been using. You don't _need_ to change anything. 👍

Reanimated provides a more comprehensive, low level abstraction for the Animated library API, giving you much greater flexibility, control and performance. Combine it with [react-native-gesture-handler](https://github.com/kmagiera/react-native-gesture-handler) for performant gesture-based interactions.

## Getting started

> Before you get started you should definitely familiarize yourself with the original [Animated API](https://facebook.github.io/react-native/docs/animated.html) first. It will do you well to be comfortable with how animations are generally done in `Animated`. (Fun Fact: Reanimated is also backwards compatible with the `Animated API`. 🙌)

Reviewer: ```suggestion
Before you get started you should definitely familiarize yourself with the original [Animated API](https://facebook.github.io/react-native/docs/animated.html). It will do you well to be comfortable with how animations are generally done in `Animated`. (Fun Fact: Reanimated is also backwards compatible with the `Animated API`. 🙌)
```

> - Refer to the [Reanimated vs. Animated](#Reanimated-vs.-Animated) to understand the differences between the old `Animated` API and Reanimated.

Reviewer: ```suggestion
- Refer to the [Reanimated vs. Animated](#Reanimated-vs.-Animated) to understand the differences between the React Native‘s `Animated` API and Reanimated.
```

Reviewer: Suggestion to change `old `Animated` API` to `React Native‘s `Animated` API`.

React Native's Animated library reimplemented.

- **Native Performance**: Declare your animations in JS, but have them run on the native thread! 🧙‍♂️
- **Precise Animations**: The API affords new levels of precision and detailed control of your animations. 🕹

> - **Backwards Compatible**: Use the same Animated library API from React Native that you've been using. You don't _need_ to change anything. 👍

Reviewer 2: It's not true. It's right in most cases but sometimes it's not.

Author: Yeah, you are right. I can see how the last "Backwards Compatible" wording could be misleading. What I meant more here was you can start using it right away, only refactoring to the specific Reanimated API as you go... I don't know how best to word it, but I think the backwards compatibility is something that is a key value of Reanimated and should be highlighted at the top (I know a couple users that probably wouldn't be using Reanimated if they couldn't start this way).

Was there something else in this section that you think should be changed? Do you think the "Native Performance" or "Precise Animations" text should be changed?

Author: ```suggestion
- **(mostly) Backwards Compatible**: Use the same Animated API from React Native that you've been using. You generally don't _need_ to change anything to get get started. 👍
```

Reviewer 2: 👌

> - **(mostly) Backwards Compatible**: Use the same Animated API from React Native that you've been using. You generally don't _need_ to change anything to get get started. 👍

Reviewer 3: ```suggestion
- **(mostly) Backwards Compatible**: Use the same Animated API from React Native that you've been using. You generally don't _need_ to change anything to get get started. 👍
- **(mostly) Backwards Compatible**: Use the same Animated API from React Native that you've been using. You generally don't _need_ to change anything to get started. 👍
```

> We aim to bring this project to be fully compatible with `Animated` API. We believe that the set of base nodes we have selected should make this possible to be done only by writing JS code and does not require significant changes in the native codebases. Here is a list of things that haven't yet been ported from the original version of `Animated` library.

Reviewer: ```suggestion
We aim to bring this project to be fully compatible with `Animated` API. We believe that the set of base nodes we have selected should make this possible to be done only by writing JS code and does not require significant changes in the native codebases.
```

Reviewer: Suggestion to move the “Here's a list […]” bit on top of the actual list.

All the functionality that missing elements provide in `Animated` can already be achieved with `react-native-reanimated` although a different methodology for implementing those may be required (e.g. check ["Running animations" section](#running-animations) to see how the implementation may differ).

> - [ ] using value offsets

Reviewer: ```suggestion
Here is a list of things that haven't yet been ported from the original version of `Animated` library:

- [ ] using value offsets
```

Reviewer: Related to the change above.

Passing array directly is equivalent to wrapping it with the [`block`](#block) command.

### At most once evaluation (the algorithm)

Unlike the original `Animated` library where each node could have been evaluated many times within a single frame, `react-native-reanimated` restricts each node to be evaluated at most once in a frame.
This restriction is required for nodes that have side-effects to be used (e.g. [`set`](#set) or [`startClock`](#startClock)).
When node is evaluated (e.g. in case of an [`add`](#add) node we want to get a sum of the input nodes) its value is cached. If within the next frame there are other nodes that want to use the output of that node instead of evaluating we return cached value.
This notion also helps with performance as we can try to evaluate as few nodes as expected.
The current algorithm for making decisions of which nodes to evaluate works as follows:

> 1. for each frame we first analyze the generated events (e.g. touch stream). It is possible that events may update some animated values.

Reviewer: ```suggestion
1. For each frame we first analyze the generated events (e.g. touch stream). It is possible that events may update some animated values.
```

Reviewer: Minor change from lowercase to uppercase to keep list consistent.

  }
}
```

This API gives the possibility to use animation with original `Animated` API. It's also a way of running animation on some interaction without necessity or rerendering view.

> 

Reviewer: ```suggestion

## Transitions 🆕

Transitions is an experimental API distributed as a part of reanimated which serves the purpose of animating between two states of view hierarchy. It is conceptually similar to `LayoutAnimation` concept from react native but gives much better control of what and how is going to be animated.

Transitions API consists of two main building blocks. First one being `Transitioning.View` which is an extension of regular react-native view, so you can use any `View` props you'd like. The `Transitioning.View` is a root of all the transition animations that will be happening and is used to scope the changes to its children. In order to have next transition animated you'd need to call `animateNextTransition()` on the `Transitioning.View` instance.

The second main bulding block is transition definition. Transitioning API uses JSX syntax that allows you to define how the transition animation should perform. You can use all the components from `Transition` object to combine the animation you want. Please see the below list for the documentation of transition components.

### Transition groups

The below set of components can be used to group one or more transitions. You can also nest transition groups in order to achieve desirable effects.

#### `<Transition.Together>`

Transitions nested under this component will run in parallel when the animation starts.

####`<Transition.Sequence>`

Transitions nested under this component will run in sequence in the order at which they are listed

### Transitions

Transition components can be used separately or as a part of a group. Each transition component has the following common properties you can use to configure the animation:

#### `durationMs`

The time animation takes to execute in milliseconds

#### `delayMs`

Use this is you want the animation to start delayed (value in milliseconds)

#### `interpolation`

Specify the transition timing curve. Possible values are: `linear`, `easeIn`, `easeOut`, `easeInOut`

#### `propagation`

Allows for the framework to automatically delay beggining of transitions across a set of different views depending on their position. The possible values are `top`, `bottom`, `left` and `right`. When `propagation="top"` it means that the first element that will start animating is the one that is closest to the top of `Transitioning.View` container, then the other views will be delayed by the amount which depends on their distance from the top edge.

### `<Transition.In>`

Allows to specify how views that get mounted durion animation transition get animated. In addition to the above parameters you can specify the type of animation using `type` prop. The possible values are: `fade`, `scale`, `slide-top`, `slide-bottom`, `slide-left`, `slide-right`.

### `<Transition.Out>`

Allows to specify how the framework should animate views that are being removed during transition. In addition to the above parameters you can specify the type of animation using `type` prop. The possible values are: `fade`, `scale`, `slide-top`, `slide-bottom`, `slide-left`, `slide-right`.

### `<Transition.Change>`

Use `Transition.Change` component to specify how components' which properties get changed during transition should be animated. The framework currently supports animating position, bounds and transforms.

### How to use it

This API is still experimental and is a subject to change. Please refer to our [Example app](https://github.com/kmagiera/react-native-reanimated/tree/master/Example/transitions) to see how it can be used in practice in the current shape.

---

```
